### PR TITLE
[mkbundle] Fix mapping length

### PR DESCRIPTION
**Pull request by jaykrell**
When I tried it, I got a different error.

The size of the mapping is the maximum file offset to map.
See https://docs.microsoft.com/en-us/windows/desktop/Memory/creating-a-file-mapping-object.

If in the debugger you simulate this fix, it gets further.

Here is a small program to exercise and see the behavior:

```
// include <windows.h>
// include <stdio.h>
// include <stdlib.h>
//
// int main()
// {
//  // Create a 128K file.
//
//  HANDLE file = CreateFile("1", GENERIC_READ | GENERIC_WRITE, 0, 0, CREATE_ALWAYS, 0, 0);
//  printf("createfile %p %d\n", file, GetLastError());
//
//  HANDLE map = CreateFileMapping(file, 0, PAGE_READWRITE, 0, 1 << 17, 0);
//  printf("createmap %p %d\n", map, GetLastError());
//
//  CloseHandle(map);
//  CloseHandle(file);
//  system("dir");
//
//  // Attempt to map the second 64K of the 128K file, with a 64K mapping.
//
//  file = CreateFile("1", GENERIC_READ | GENERIC_WRITE, 0, 0, OPEN_EXISTING, 0, 0);
//  printf("createfile %p %d\n", file, GetLastError());
//
//  map = CreateFileMapping(file, 0, PAGE_READWRITE, 0, 1 << 16, 0);
//  printf("createmap %p %d\n", map, GetLastError());
//
//   // This fails.
//
//  PVOID view = MapViewOfFile(map, FILE_MAP_WRITE, 0, 1 << 16, 1 << 16);
//  printf("view %p %d\n", view, GetLastError());
//
//  CloseHandle(map);
//
//  // Use a 128K mapping instead. It works.
//
//  map = CreateFileMapping(file, 0, PAGE_READWRITE, 0, 1 << 17, 0);
//  printf("createmap %p %d\n", map, GetLastError());
//
//  view = MapViewOfFile(map, FILE_MAP_WRITE, 0, 1 << 16, 1 << 16);
//  printf("view %p %d\n", view, GetLastError());
// }
```

Here is debugging mono and simulating the fix in the debugger (I did not build or test
the fix):

```

\bin\amd64\cdb C:\Users\jaykrell\Downloads\output\output.exe

0:000> bm kernel*!*createfilemap*
0:000> g

Breakpoint 2 hit
KERNELBASE!CreateFileMappingW:
00007ffd`0a449de0 4883ec48        sub     rsp,48h
0:000> dv
                  hFile = 0x00000000`00000200
lpFileMappingAttributes = 0x00000000`00000000
              flProtect = 2
      dwMaximumSizeHigh = 0
       dwMaximumSizeLow = 0x477000
                 lpName = 0x00000000`00000000 ""

0:000> bm kern*!*mapviewoffile*
0:000> g
Breakpoint 38 hit
KERNEL32!MapViewOfFileStub:
00007ffd`0b09bfa0 48ff2591c10500  jmp     qword ptr [KERNEL32!_imp_MapViewOfFile (00007ffd`0b0f8138)] ds:00007ffd`0b0f8138={KERNELBASE!MapViewOfFile (00007ffd`0a465ad0)}
0:000> dv
  hFileMappingObject = 0x00000000`00000204
     dwDesiredAccess = 4
    dwFileOffsetHigh = 0
     dwFileOffsetLow = 0x30000
dwNumberOfBytesToMap = 0x477000
0:000> gu
0:000> r rax
rax=0000000000000000
0:000> !gle
LastErrorValue: (Win32) 0x5 (5) - Access is denied.
LastStatusValue: (NTSTATUS) 0xc000001f - {Invalid Mapping}  An attempt was made to create a view for a section which is bigger than the section.

0:000> .restart
0:000> bm kernel*!*createfilemap*
0:000> g
0:000> dv
                  hFile = 0x00000000`00000220
lpFileMappingAttributes = 0x00000000`00000000
              flProtect = 2
      dwMaximumSizeHigh = 0
       dwMaximumSizeLow = 0x477000
                 lpName = 0x00000000`00000000 ""
0:000> ?? dwMaximumSizeLow += 0x30000
unsigned long 0x4a7000
0:000> bc *
0:000> g
Corlib not in sync with this runtime: The runtime did not find the mscorlib.dll it expected. Expected interface version CA4932AE-2294-4ECD-B863-BF98FDD84F33 but found 0CC970F6-6F25-4218-9427-3E4C3DC8DE33. Check that your runtime and class libraries are matching.
Loaded from: mscorlib.dll
Download a newer corlib or a newer runtime at http://www.mono-project.com/download.

```

Furthermore, it can be reasoned, that if this was the maximum view size,
it would be one DWORD parameter originally, later redefined as SIZE_T.
There'd be no need for 64bits of size on a 32bit system.

Fixes https://github.com/mono/mono/issues/11643

### CODE REVIEW

**Comment by lateralusX** [comment]
Where you able to successfully run the test case as outlined in the issue with this fix?

**Comment by jaykrell** [comment]
I reproed a different problem with the binary, simulated the fix in the debugger by poking parameters.
Building mono on Windows I did not do.
Also the small test I shows I compiled and ran.
The evidence adds up.

**Comment by jaykrell** [comment]
The reported error message seems impossible in the current code. It should be longer.

**Comment by lateralusX** [comment]
OK, I will apply the fix local and build and test on Windows before approving this.

**Comment by jaykrell** [comment]
https://docs.microsoft.com/en-us/windows/desktop/Memory/creating-a-file-mapping-object

Explains it.

The size of the file mapping object that you select controls how far into the file you can "see" with memory mapping. If you create a file mapping object that is 500 Kb in size, you have access only to the first 500 Kb of the file, regardless of the size of the file. Since it does not cost you any system resources to create a larger file mapping object, create a file mapping object that is the size of the file (set the dwMaximumSizeHigh and dwMaximumSizeLow parameters of CreateFileMapping both to zero) even if you do not expect to view the entire file. The cost in system resources comes in creating the views and accessing them.

If you want to view a portion of the file that does not start at the beginning of the file, you must create a file mapping object. This object is the size of the portion of the file that you want to view plus the offset into the file.

**Comment by jaykrell** [comment]
So another fix is just pass 0.

**Comment by jaykrell** [comment]
Not passing 0, in case there are uses of this to grow the file.

File: mono/utils/mono-mmap-windows.c [hunk 1]
```diff
@@ -142,25 +142,18 @@ mono_file_map_error (size_t length, int flags, int fd, guint64 offset, void **re
 	void *ptr = NULL;
 	HANDLE mapping = NULL;
 	int const prot = mono_mmap_win_prot_from_flags (flags);
-	/* translate the flags */
```
Comment (luhenry) [inline mono/utils/mono-mmap-windows.c]: Revert that removal.

File: mono/utils/mono-mmap-windows.c [hunk 2]
```diff
@@ -143,24 +143,18 @@ mono_file_map_error (size_t length, int flags, int fd, guint64 offset, void **re
 	HANDLE mapping = NULL;
 	int const prot = mono_mmap_win_prot_from_flags (flags);
 	/* translate the flags */
-	/*if (flags & MONO_MMAP_PRIVATE)
```
Comment (luhenry) [inline mono/utils/mono-mmap-windows.c]: Revert that removal.

I have to admit https://github.com/mono/mono/pull/13061/commits/4a689bf15de06ba4c8394cd7a6a5a46ba062fbe2 is hilarious! I mean, seriously? That's a beautiful way to bend the review to its maximum, congrats! :)

File: mono/utils/mono-mmap-windows.c (hunk 1 above)
Comment (jaykrell) [reply mono/utils/mono-mmap-windows.c]: I restored the first line, which remained accurate. I kept removed the dead code. PR feedback is ambiguous to me as to if the entire removal or just the first part was referred to.

File: mono/utils/mono-mmap-windows.c (hunk 2 above)
Comment (jaykrell) [reply mono/utils/mono-mmap-windows.c]: It was ambiguous. I did accidentally remove a comment along with dead code. Dead code now restored.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by lateralusX** [comment]
Successfully verified fix on local build.

**Comment by lateralusX** [comment]
Backport?

**Pull request merged**